### PR TITLE
config: split out DeltaSubscriptionState

### PR DESCRIPTION
**Pull request by fredlas**
Split out about half of DeltaSubscriptionImpl into DeltaSubscriptionState, one of which is owned by DeltaSubscriptionImpl. The upcoming aggregated implementation will aggregate-ize the remaining code of DeltaSubscriptionImpl (in a new class with a new name), and replace the current single DeltaSubscriptionState instance with a map of them.

Risk Level: low
Testing: existing delta xDS unit test is almost entirely for the split out logic

#4991 #5270

### CODE REVIEW

**Comment by fredlas** [comment]
Exactly one test in subscription_impl_test is failing on delta, because it expects a combined subscription update + ACK where the new code sends an ACK first. I'm not sure how to handle that one other than disabling it for delta. Is that ok to do? I think the multi-xDS-implementation test harness setup might be a little too fine grained, i.e. expecting explicit sequences of mechanical events, rather than meaningful end results.

Also, this might be a good PR to split delta_subscription_impl.h into a .cc, now that there are no more templates. Or I can just do that in a followup.

**Comment by zuercher** [comment]
If the test is covering a case that cannot happen with delta subscriptions, then I think it's fine to disable for that case.

**Comment by fredlas** [comment]
Thanks, "fine to disable" is always nice to hear :)

I realized the "inflexibility" is actually in the test harness, so I could just go fix it for delta.

File: source/common/config/delta_subscription_impl.h [hunk 1]
```diff
@@ -261,83 +113,50 @@ class DeltaSubscriptionImpl : public Subscription,
...
+  // still send a request; it just won't have an ACK.
+  //
+  // More details: DeltaDiscoveryRequest plays two independent roles:
+  // 1) informing the server of what resources we're interested in, and
+  // 2) acknowledging resources the server has sent us.
+  // Each entry in this queue was added for exactly one of those purposes, but since the
+  // subscription interest is tracked separately, in a non-queue way, subscription changes can get
+  // mixed in with an ACK request. In that case, the entry that the subscription change originally
+  // queued up *does* still get sent, just empty and pointless. (TODO(fredlas) we would like to skip
+  // those no-op requests).
```
Comment (htuch) [inline source/common/config/delta_subscription_impl.h]: Yeah, I'm still regretting the choice made early in incremental of keeping state-of-the-world's approach to mixing ACK and subscription change. Even just trying to think through all the corner cases and interleavings here is blowing my mind. Do you reckon it's possible we can make this distinction explicit without causing too much implementation churn?

File: source/common/config/delta_subscription_state.cc [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  resource_versions_[resource_name] = ResourceVersion(resource_version);
+  resource_names_.insert(resource_name);
+}
+
+void DeltaSubscriptionState::setResourceWaitingForServer(const std::string& resource_name) {
+  resource_versions_[resource_name] = ResourceVersion();
+  resource_names_.insert(resource_name);
+}
+
+void DeltaSubscriptionState::lostInterestInResource(const std::string& resource_name) {
```
Comment (htuch) [inline source/common/config/delta_subscription_state.cc]: Nit: prefer `setLostInterestInResource`

File: test/test_common/utility_test.cc [hunk 3]
```diff
@@ -27,6 +29,78 @@ TEST(headerMapEqualIgnoreOrder, NotEqual) {
   EXPECT_FALSE(TestUtility::headerMapEqualIgnoreOrder(lhs, rhs));
 }
 
+TEST(protoEqIgnoreField, ActuallyEqual) {
```
Comment (htuch) [inline test/test_common/utility_test.cc]: Nit: `ProtoEqIgnoreField`

File: test/test_common/utility.h [hunk 4]
```diff
@@ -570,6 +580,28 @@ MATCHER_P(ProtoEq, expected, "") {
...
 
+MATCHER_P2(ProtoEqIgnoringField, expected, ignored_field, "") {
+  const bool equal = TestUtility::protoEqualIgnoringField(arg, expected, ignored_field);
+  if (!equal) {
+    int line_fill_len =
+        (68 - (std::string(ignored_field).length() + std::string("(but ignoring )").length()));
+    int first_half_len = line_fill_len / 2;
+    int second_half_len = line_fill_len - first_half_len;
+    std::string first_half(first_half_len, '=');
+    std::string second_half(second_half_len, '=');
```
Comment (htuch) [inline test/test_common/utility.h]: Do we need this ASCII art? It's OK if it helps make things much clearer, but the fewer lines dedicated to drawing ASCII art when trying to debug/understand code the better IMHO.

File: source/common/config/delta_subscription_impl.h [hunk 5]
```diff
@@ -44,210 +39,67 @@ class DeltaSubscriptionImpl : public Subscription,
...
-      // that, it's difficult to distinguish remove-add-remove from add-remove (because "remove-add"
-      // has to be treated as equivalent to just "add").
-      names_added_.erase(r);
-      names_removed_.insert(r);
-    }
-
-    stats_.update_attempt_.inc();
     // Tell the server about our new interests (but only if there are any).
-    if (!names_added_.empty() || !names_removed_.empty()) {
+    ASSERT(state_, "Must call start() on DeltaSubscriptionImpl before doing anything with it.");
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: Done.

File: source/common/config/delta_subscription_impl.h [hunk 6]
```diff
@@ -44,210 +39,67 @@ class DeltaSubscriptionImpl : public Subscription,
...
-        (*request_.mutable_initial_resource_versions())[resource.first] = resource.second.version();
-      }
-      // As mentioned above, fill resource_names_subscribe with everything.
-      names_added_.insert(resource.first);
-    }
-    names_removed_.clear();
-    request_.set_type_url(type_url_);
-    request_.mutable_node()->MergeFrom(local_info_.node());
+    ASSERT(state_, "Must call start() on DeltaSubscriptionImpl before doing anything with it.");
+    state_->set_first_request_of_new_stream(true);
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: While this is an instance of a thing with states like that, and a state enum is indeed almost always the right way to go, I think (unless/until things get more complex) this case is simpler with just this bool. There is one other boolean state, `paused_`, but that is used/modified completely independently from `first_request_of_new_stream_`. I think trying to combine them into an enum would result in needing convenience functions to interpret the enum values back to bool, with the enum values themselves never actually getting used.

File: source/common/config/delta_subscription_state.h [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  DeltaSubscriptionState(const std::string& type_url, const std::set<std::string>& resource_names,
+                         SubscriptionCallbacks& callbacks, const LocalInfo::LocalInfo& local_info,
+                         std::chrono::milliseconds init_fetch_timeout,
+                         Event::Dispatcher& dispatcher, SubscriptionStats& stats);
+
+  void setInitFetchTimeout(Event::Dispatcher& dispatcher);
+
+  void pause();
+  void resume();
+  bool paused() { return paused_; }
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_state.h]: Done.

File: source/common/config/delta_subscription_impl.h (hunk 1 above)
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: I think the `oneof` approach that we've briefly discussed wouldn't disrupt this code much at all.

Can you clarify for me what aspects of xDS need to be involved in the dependency ordering? Is it just the server's resources and the Envoy's ACKs, or do the Envoy's subscription changes also sometimes need to be ordered with its ACKs in a certain way?

If the subscription changes don't need to be ordered with ACKs, then the two mechanisms seem pretty independent from each other. In that case, this mixing wouldn't actually feel very uncomfortable to me at the conceptual protocol level.

As for the specific implementation here, I definitely do think it's hard to follow, but can be cleaned up. These nullopt queue entries are ultimately a hack to allow subscription changes to use the ACK queue as a way to signal that they need to be sent. There is a cleaner way. The queue can be for ACKs *only* - no longer absl::optional - and the DeltaSubscriptionState can have a `bool subscriptionUpdatePending()`. Then, rather than the sending loop's condition being `!ack_queue_.empty()`, it would be `!ack_queue_.empty() || state_->subscriptionUpdatePending()`. What do you think?

(Note, the above paragraph, as well as the current implementation, would be outright wrong if subscription changes do need to be ordered with ACKs).

File: source/common/config/delta_subscription_state.cc (hunk 2 above)
Comment (fredlas) [reply source/common/config/delta_subscription_state.cc]: Good catch, with the past tense this name is very misleading, sounding like a const getter.

File: test/test_common/utility_test.cc (hunk 3 above)
Comment (fredlas) [reply test/test_common/utility_test.cc]: Done, and changed the others in the file as well.

File: test/test_common/utility.h (hunk 4 above)
Comment (fredlas) [reply test/test_common/utility.h]: It's
```
==========================Expected proto:===========================
===================(but ignoring response_nonce)====================
type_url: "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment"
response_nonce: "nonce"
------------------is not equal to actual proto:---------------------
type_url: "type.googleapis.com/envoy.api.v2.Listener"
response_nonce: "nonce"
====================================================================
```
vs
```
==========================Expected proto:===========================
=====(but ignoring response_nonce)=====
type_url: "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment"
response_nonce: "nonce"
------------------is not equal to actual proto:---------------------
type_url: "type.googleapis.com/envoy.api.v2.Listener"
response_nonce: "nonce"
====================================================================
```
I think it's reasonably readable but shabby looking the simpler way. (Not having any =s would in fact make it unreadable, I think). I split that code out into a helper; maybe it's cleaner this way? Actually, this new helper could be used in all of these matcher debug strings now, although I think it's clearer to stick with the hardcoded strings where possible.

File: source/common/config/delta_subscription_impl.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/config/delta_subscription_impl.h]: OK; can we then make this `startStream()` or something? I.e. there is no reason to set this `false` in the interface.

File: source/common/config/delta_subscription_impl.h (hunk 1 above)
Comment (htuch) [reply source/common/config/delta_subscription_impl.h]: It would be really easy to reason about if we maintained these properties:
* Each message sent by the server has an ACK/NACK reply at some point in the future; it would be reasonable for the server to wait for this. It could issue a bunch of messages and then wait for their ACKS/NACKs to come back.
* Independent of ^^, Envoy will send a series of subscription changes as needed. These relate in no way to the ACK/NACKs and reflect the client intention.

Does it sound reasonable to make that work from a code perspective?

File: test/test_common/utility.h (hunk 4 above)
Comment (htuch) [reply test/test_common/utility.h]: OK, if you want to make this a utility shared by all that's fine.

File: source/common/config/delta_subscription_impl.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: Ah yes, I see what you mean. Thanks, much cleaner this way. I also renamed the variable itself, maybe slightly more descriptive.

File: source/common/config/delta_subscription_impl.h (hunk 1 above)
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: Excellent, thank you. I believe I now have something that takes advantage of that independence to keep things clear, while also still being able to mix them into the same protobuf.

(Also, note the no-op expectSendMessage()s deleted from the test! Hooray!)

File: test/test_common/utility.h (hunk 4 above)
Comment (fredlas) [reply test/test_common/utility.h]: Done.

Comment (fredlas) [reply test/test_common/utility.h]: Unfortunately, with the helper used everywhere, clang format has decided to make it a little hideous.

File: source/common/config/delta_subscription_state.cc [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+// Not having sent any requests yet counts as an "update pending" since you're supposed to resend
+// the entirety of your interest at the start of a stream, even if nothing has changed.
+bool DeltaSubscriptionState::subscriptionUpdatePending() const {
+  return !names_added_.empty() || !names_removed_.empty() ||
+         !any_request_sent_yet_in_current_stream_;
+}
+
+UpdateAck
+DeltaSubscriptionState::handleResponse(const envoy::api::v2::DeltaDiscoveryResponse& message) {
```
Comment (fredlas) [inline source/common/config/delta_subscription_state.cc]: Note: I haven't been considering this to be particularly callbacky, other than the fact that it's always called by a callback (onDiscoveryResponse). I think of it more as "update the state machine with this new input". If you think it makes more sense with the on..... nomenclature, though, I am fine with changing it.

Comment (fredlas) [reply source/common/config/delta_subscription_state.cc]: Sorry, thinking a little more, I realize I actually want DeltaSubscriptionState to be completely walled off from anything like callbacks. It should just be its own little chunk of state, with functions that modify that state and produce output. Maybe these should get even *less* callbacky names.

Comment (htuch) [reply source/common/config/delta_subscription_state.cc]: This is fine, these are private methods, so all good.

File: source/common/config/delta_subscription_state.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
+#pragma once
```
Comment (htuch) [inline source/common/config/delta_subscription_state.h]: Arguably we should also add unit tests for `DeltaSubscriptionState` rather than relying on testing via the parent module, but we have full coverage, so consider this an optional improvement for the future as it proves useful.

File: source/common/config/delta_subscription_impl.h [hunk 10]
```diff
@@ -44,210 +39,55 @@ class DeltaSubscriptionImpl : public Subscription,
...
-
-  bool shouldSendDiscoveryRequest() {
-    if (paused_) {
+  // Checks whether external conditions allow sending a DeltaDiscoveryRequest. (Does not check
+  // whether we *want* to send a DeltaDiscoveryRequest).
+  bool canSendDiscoveryRequest() {
+    if (state_->paused()) {
       ENVOY_LOG(trace, "API {} paused; discovery request on hold for now.", type_url_);
       return false;
     } else if (!grpc_stream_.grpcStreamAvailable()) {
```
Comment (htuch) [inline source/common/config/delta_subscription_impl.h]: We're missing coverage here. We were before, but we still are :) Can you add this in a future PR?

File: source/common/config/delta_subscription_impl.h [hunk 11]
```diff
@@ -260,84 +100,55 @@ class DeltaSubscriptionImpl : public Subscription,
...
-
-    // If true, we currently have no version of this resource - we are waiting for the server to
-    // provide us with one.
-    bool waitingForServer() const { return version_ == absl::nullopt; }
-    // Must not be called if waitingForServer() == true.
-    std::string version() const {
-      ASSERT(version_.has_value());
-      return version_.value_or("");
+  void trySendDiscoveryRequests() {
+    while (wantToSendDiscoveryRequest() && canSendDiscoveryRequest()) {
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: :D thanks! I was also very happy when I realized it could be phrased like this.

File: source/common/config/delta_subscription_state.h (hunk 9 above)
Comment (fredlas) [reply source/common/config/delta_subscription_state.h]: Will do, and this test work will be easily parallelizable with the other ongoing xDS work, so that's good.

File: source/common/config/delta_subscription_impl.h (hunk 10 above)
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: Will do, and this test work will be easily parallelizable with the other ongoing xDS work, so that's good.

File: source/common/config/delta_subscription_state.h (hunk 9 above)
Comment (fredlas) [reply source/common/config/delta_subscription_state.h]: Done.

File: source/common/config/delta_subscription_impl.h (hunk 10 above)
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: Done.
